File: ai-js/src/agent/base-agent.ts
import { Logger } from '@nestjs/common';
import { AgentState } from './model/agent-state.enum';
import { ChatModel } from '../llm/chat-model';
import { createUserMessage, Message } from './model/message';
import { Observable, Subject } from 'rxjs';
import { EventName } from './model/event-name.enum';

export abstract class BaseAgent {
  protected readonly logger = new Logger(
    // BaseAgent.name
    this.constructor.name,
  );

  name = this.constructor.name;
  systemPrompt = '';
  nextStepPrompt = '';
  state: AgentState = AgentState.IDLE;
  currentStep = 0;
  maxSteps = 10;
  chatModel: ChatModel | null = null;
  messageList: Message[] = [];

  protected get isNotFinished() {
    return this.state !== AgentState.FINISHED;
  }

  async run(userPrompt: string): Promise<string> {
    if (this.state !== AgentState.IDLE) {
      throw new Error(`Cannot run agent from state: ${this.state}`);
    }
    if (!userPrompt.trim()) {
      throw new Error('Cannot run agent with empty user prompt');
    }
    this.state = AgentState.RUNNING;
    this.messageList.push(createUserMessage(userPrompt));
    const results: string[] = [];
    try {
      for (let i = 0; i < this.maxSteps && this.isNotFinished; i++) {
        const stepNumber = i + 1;
        this.currentStep = stepNumber;
        this.logger.log(`Executing step ${stepNumber}/${this.maxSteps}`);
        const stepResult = await this.step();
        const result = `Step ${stepNumber}: ${stepResult}`;
        results.push(result);
      }
      return results.join('\n');
    } catch (err) {
      this.state = AgentState.ERROR;
      this.logger.error('Executing agent error:', err);
      return 'Executing agent error';
    } finally {
      this.cleanup();
    }
  }

  runStream(userPrompt: string): Observable<string> {
    const subject = new Subject<string>();

    // Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the `void` operator.
    (async () => {
      try {
        if (this.state !== AgentState.IDLE) {
          subject.next(
            JSON.stringify({
              type: EventName.ERROR,
              content: `Cannot run agent from state: ${this.state}`,
            }),
          );
          subject.complete();
          return;
        }
        if (!userPrompt.trim()) {
          subject.next(
            JSON.stringify({
              type: EventName.ERROR,
              content: 'Cannot run agent with empty user prompt',
            }),
          );
          subject.complete();
          return;
        }
      } catch (err) {
        subject.error(err);
        return;
      }

      this.state = AgentState.RUNNING;
      this.messageList.push(createUserMessage(userPrompt));
      try {
        for (let i = 0; i < this.maxSteps && this.isNotFinished; i++) {
          const stepNumber = i + 1;
          this.currentStep = stepNumber;
          this.logger.log(`Executing step ${stepNumber}/${this.maxSteps}`);
          subject.next(
            JSON.stringify({
              type: EventName.STEP_START,
              step: stepNumber,
              max_steps: this.maxSteps,
            }),
          );
          const stepResult = await this.stepWithEmitter(subject);
          subject.next(
            JSON.stringify({
              type: EventName.STEP_RESULT,
              step: stepNumber,
              content: stepResult,
            }),
          );
        }
        if (this.currentStep >= this.maxSteps) {
          this.state = AgentState.FINISHED;
          subject.next(
            JSON.stringify({
              type: AgentState.FINISHED,
              content: `Terminated, reached max steps (${this.maxSteps})`,
            }),
          );
        }
        subject.complete();
      } catch (err) {
        this.state = AgentState.ERROR;
        this.logger.error('Executing agent error:', err);
        subject.next(
          JSON.stringify({
            type: EventName.ERROR,
            content: 'Executing agent error',
          }),
        );
        subject.complete();
      } finally {
        this.cleanup();
      }
    })().catch((err) => {
      this.logger.error('Stream error:', err);
    });

    return subject.asObservable();
  }

  // eslint-disable-next-line @typescript-eslint/no-unused-vars
  protected async stepWithEmitter(emitter: Subject<string>): Promise<string> {
    return this.step();
  }

  abstract step(): Promise<string>;

  protected cleanup(): void {}
}

File: src/main/java/com/yupi/yuaiagent/agent/BaseAgent.java
package com.yupi.yuaiagent.agent;

import cn.hutool.core.util.StrUtil;
import cn.hutool.json.JSONUtil;
import com.yupi.yuaiagent.agent.model.AgentState;
import java.io.IOException;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;
import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ai.chat.client.ChatClient;
import org.springframework.ai.chat.messages.Message;
import org.springframework.ai.chat.messages.UserMessage;
import org.springframework.web.servlet.mvc.method.annotation.SseEmitter;

/**
 * 抽象基础代理类, 用于管理代理状态和执行流程;
 *
 * <p>提供状态转换、内存管理和基于步骤的执行循环的基础功能; 子类必须实现step方法;
 */
@Data
@Slf4j
public abstract class BaseAgent {

  // 核心属性
  private String name;

  // 提示词
  private String systemPrompt;
  private String nextStepPrompt;

  // 代理状态
  private AgentState state = AgentState.IDLE;

  // 执行步骤控制
  private int currentStep = 0;
  private int maxSteps = 10;

  // LLM 大模型
  private ChatClient chatClient;

  // Memory 记忆 (需要自主维护会话上下文)
  private List<Message> messageList = new ArrayList<>();

  // SSE 发射器 (用于流式输出)
  protected SseEmitter sseEmitter;

  /** 发送 SSE 结构化消息 */
  protected void sendSseMessage(String type, Map<String, Object> data) {
    if (sseEmitter != null) {
      try {
        Map<String, Object> message = new HashMap<>();
        message.put("type", type);
        message.putAll(data);
        sseEmitter.send(JSONUtil.toJsonStr(message));
      } catch (IOException e) {
        log.error("Failed to send SSE message", e);
      }
    }
  }

  /** 发送简单的 SSE 消息 */
  protected void sendSseMessage(String type, String content) {
    Map<String, Object> data = new HashMap<>();
    data.put("content", content);
    sendSseMessage(type, data);
  }

  /**
   * 运行代理
   *
   * @param userPrompt 用户提示词
   * @return 执行结果
   */
  public String run(String userPrompt) {
    // 1、基础校验
    if (this.state != AgentState.IDLE) {
      throw new RuntimeException("Cannot run agent from state: " + this.state);
    }
    if (StrUtil.isBlank(userPrompt)) {
      throw new RuntimeException("Cannot run agent with empty user prompt");
    }
    // 2、执行, 更改状态
    this.state = AgentState.RUNNING;
    // 记录消息上下文
    messageList.add(new UserMessage(userPrompt));
    // 保存结果列表
    List<String> results = new ArrayList<>();
    try {
      // 执行循环
      for (int i = 0; i < maxSteps && state != AgentState.FINISHED; i++) {
        int stepNumber = i + 1;
        currentStep = stepNumber;
        log.info("Executing step {}/{}", stepNumber, maxSteps);
        // 单步执行
        String stepResult = step();
        String result = "Step " + stepNumber + ": " + stepResult;
        results.add(result);
      }
      // 检查是否超出步骤限制
      if (currentStep >= maxSteps) {
        state = AgentState.FINISHED;
        results.add("Terminated: Reached max steps (" + maxSteps + ")");
      }
      return String.join("\n", results);
    } catch (Exception e) {
      state = AgentState.ERROR;
      log.error("error executing agent", e);
      return "执行错误" + e.getMessage();
    } finally {
      // 3、清理资源
      this.cleanup();
    }
  }

  /**
   * 运行代理 (流式输出)
   *
   * @param userPrompt 用户提示词
   * @return 执行结果
   */
  public SseEmitter runStream(String userPrompt) {
    // 创建一个超时时间较长的 SseEmitter
    SseEmitter emitter = new SseEmitter(300000L); // 5 分钟超时
    this.sseEmitter = emitter;

    // 使用线程异步处理, 避免阻塞主线程
    CompletableFuture.runAsync(
        () -> {
          // 1、基础校验
          try {
            if (this.state != AgentState.IDLE) {
              sendSseMessage("error", "无法从状态运行代理：" + this.state);
              emitter.complete();
              return;
            }
            if (StrUtil.isBlank(userPrompt)) {
              sendSseMessage("error", "不能使用空提示词运行代理");
              emitter.complete();
              return;
            }
          } catch (Exception e) {
            emitter.completeWithError(e);
          }
          // 2、执行, 更改状态
          this.state = AgentState.RUNNING;
          // 记录消息上下文
          messageList.add(new UserMessage(userPrompt));
          // 保存结果列表
          List<String> results = new ArrayList<>();
          try {
            // 执行循环
            for (int i = 0; i < maxSteps && state != AgentState.FINISHED; i++) {
              int stepNumber = i + 1;
              currentStep = stepNumber;
              log.info("Executing step {}/{}", stepNumber, maxSteps);

              // 发送步骤开始事件
              Map<String, Object> stepStartData = new HashMap<>();
              stepStartData.put("step", stepNumber);
              stepStartData.put("maxSteps", maxSteps);
              sendSseMessage("step_start", stepStartData);

              // 单步执行
              String stepResult = step();
              String result = "Step " + stepNumber + ": " + stepResult;
              results.add(result);

              // 发送步骤结果
              Map<String, Object> stepResultData = new HashMap<>();
              stepResultData.put("step", stepNumber);
              stepResultData.put("content", stepResult);
              sendSseMessage("step_result", stepResultData);
            }
            // 检查是否超出步骤限制
            if (currentStep >= maxSteps) {
              state = AgentState.FINISHED;
              results.add("Terminated: Reached max steps (" + maxSteps + ")");
              sendSseMessage("finished", "达到最大步骤 (" + maxSteps + ")");
            }
            // 正常完成
            emitter.complete();
          } catch (Exception e) {
            state = AgentState.ERROR;
            log.error("error executing agent", e);
            try {
              sendSseMessage("error", "执行错误：" + e.getMessage());
              emitter.complete();
            } catch (Exception ex) {
              emitter.completeWithError(ex);
            }
          } finally {
            // 3、清理资源
            this.sseEmitter = null;
            this.cleanup();
          }
        });

    // 设置超时回调
    emitter.onTimeout(
        () -> {
          this.state = AgentState.ERROR;
          this.sseEmitter = null;
          this.cleanup();
          log.warn("SSE connection timeout");
        });
    // 设置完成回调
    emitter.onCompletion(
        () -> {
          if (this.state == AgentState.RUNNING) {
            this.state = AgentState.FINISHED;
          }
          this.sseEmitter = null;
          this.cleanup();
          log.info("SSE connection completed");
        });
    return emitter;
  }

  /**
   * 定义单个步骤
   *
   * @return
   */
  public abstract String step();

  /** 清理资源 */
  protected void cleanup() {
    // 子类可以重写此方法来清理资源
  }
}
